### PR TITLE
make sure that our error handling runs with the GIL enabled

### LINKED ISSUE #92684: Crash when using CUDA Graphs and calling
### 🐛 Describe the bug

Using CUDA graphs crashes PyTorch with no raised exception. This is the code to reproduce it.

```python
import torch

x = torch.randn(3, 4).cuda()
mask = x.ge(0.5).cuda()

graph = torch.cuda.graphs.CUDAGraph()

with torch.cuda.graph(graph):
    d = x.masked_select(mask)
```

If I attach with GDB I get a crash on `C10_CUDA_CHECK(cudaStreamSynchronize(stream));` that the operation is not supported during capture.

https://github.com/pytorch/pytorch/blob/b2ca2c86624d0df62b59e5c42a37457afb2d5145/c10/cuda/CUDAFunctions.h#L82

### Versions

Collecting environment information...
PyTorch version: 1.14.0 

This is a local release build of PyTorch @ `ce9963e6ba0e40b8307477f5b4113733e7a30ec2`

Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 14 2022, 12:59:47)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.13.0-1023-aws-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 520.61.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.0
[pip3] open-clip-torch==2.8.2
[pip3] pytorch-lightning==1.8.5.post0
[pip3] rotary-embedding-torch==0.1.5
[pip3] torch==1.14.0
[pip3] torchdata==0.5.1
[pip3] torchmetrics==0.11.0
[pip3] torchvision==0.15.0a0+46b7e27
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @mcarilli

### PR DESCRIPTION
**Pull request by albanD**
Fixes https://github.com/pytorch/pytorch/issues/92684

I checked the other use case of this API and they never release the GIL

### CODE REVIEW

**Comment by albanD** [comment]
Ho the version of clangformat I used is too old. Someone updated it haha

File: test/test_cuda.py [hunk 1]
```diff
@@ -3334,6 +3335,40 @@ def test_graph_capture_simple(self):
...
+exit(2)
+"""
+        try:
+            a = subprocess.check_output(
+                [sys.executable, '-c', script],
+                stderr=subprocess.STDOUT,
+                # On Windows, opening the subprocess with the default CWD makes `import torch`
+                # fail, so just set CWD to this script's directory
+                cwd=os.path.dirname(os.path.realpath(__file__)),)
+            print(a)
```
Comment (ngimel) [inline test/test_cuda.py:3364]: stray prints?

**Review by ngimel — APPROVED** [review]
How do other pybinded functions behave now wrt gil?

File: test/test_cuda.py (hunk 1 above)
Comment (albanD) [reply test/test_cuda.py:3364]: Oops

**Comment by albanD** [comment]
> How do other pybinded functions behave now wrt gil?

This PR is only changing the behavior of the functions touched in the PR (making sure we have the GIL when doing the error handling).
Other functions are not changed.

But in general, by default the GIL is not released and the user has to do it manually in their code if they want to.

**Pull request closed without merging**